src/game/logic/map/polygontrigger.cpp

```diff
@@ -21,28 +24,417 @@
+            maxTriggerId = id;
+        }
+
+        for (int i = 0; i < numPoints; i++) {
+            ICoord3D loc;
+            loc.x = file.Read_Int32();
+            loc.y = file.Read_Int32();
+            loc.z = file.Read_Int32();
+            trigger->Add_Point(loc);
+        }
+
+        if (numPoints >= 2) {
+            if (pPrevTrig) {
+                pPrevTrig->Set_Next(trigger);
+            } else {
+                Add_Polygon_Trigger(trigger);
+            }
+
+            pPrevTrig = trigger;
+        } else {
+            captainslog_debug("Deleting polygon trigger '%s' with %d points.\n", triggerName.Str(), numPoints);
+            trigger->Delete_Instance();
+        }
+    }
+
+    if (info->version == 1) {
+        PolygonTrigger *trigger = NEW_POOL_OBJ(PolygonTrigger, 4);
+        trigger->Set_Water_Area(true);
+        trigger->m_triggerID = maxTriggerId;
+        maxTriggerId++;
+
+        ICoord3D loc;
+        loc.x = -300;
+        loc.y = -300;
+        loc.z = 7;
+        trigger->Add_Point(loc);
+
+        loc.x = (int)(g_theWriteableGlobalData->m_waterExtentX + 300.0f);
+        trigger->Add_Point(loc);
+
+        loc.y = (int)(g_theWriteableGlobalData->m_waterExtentY + 300.0f);
+        trigger->Add_Point(loc);
+
+        loc.x = -300;
+        trigger->Add_Point(loc);
+
+        if (pPrevTrig) {
+            pPrevTrig->Set_Next(trigger);
+        } else {
+            Add_Polygon_Trigger(trigger);
+        }
+        pPrevTrig = trigger;
+    }
+
+    s_currentID = maxTriggerId + 1;
+    captainslog_dbgassert(file.At_End_Of_Chunk(), "Incorrect data file length.");
+    return true;
 }
 
 void PolygonTrigger::Delete_Triggers()
 {
-#ifdef GAME_DLL
-    Call_Function<void>(PICK_ADDRESS(0x005714C0, 0x006CF080));
-#endif
+    PolygonTrigger *p = s_thePolygonTriggerListPtr;
+    s_thePolygonTriggerListPtr = nullptr;
+    s_currentID = 1;
+    p->Delete_Instance();
 }
 
 bool PolygonTrigger::Point_In_Trigger(ICoord3D &point) const
 {
-#ifdef GAME_DLL
-    return Call_Method<bool, const PolygonTrigger, ICoord3D &>(PICK_ADDRESS(0x00571660, 0x006CF4D4), this, point);
-#else
-    return 0;
-#endif
+    if (m_boundsNeedsUpdate) {
+        Update_Bounds();
+    }
+
+    if (point.x < m_bounds.lo.x) {
+        return false;
+    }
+
+    if (point.y < m_bounds.lo.y) {
+        return false;
+    }
+
+    if (point.x > m_bounds.hi.x) {
+        return false;
+    }
+
+    if (point.y > m_bounds.hi.y) {
+        return false;
+    }
+
+    bool b = false;
+
+    for (int i = 0; i < m_numPoints; i++) {
+        int x1 = m_points[i].x;
+        int y1 = m_points[i].y;
+        int x2;
+        int y2;
+        if (i == m_numPoints - 1) {
+            x2 = m_points[0].x;
+            y2 = m_points[0].y;
+        } else {
+            x2 = m_points[i + 1].x;
+            y2 = m_points[i + 1].y;
+        }
+
+        if (y1 != y2 && (y1 >= point.y || y2 >= point.y) && (y1 < point.y || y2 < point.y)

```

Comment:
interestingly 
https://wrf.ecse.rpi.edu/Research/Short_Notes/pnpoly.html
its that algo